src/test/java/net/satago/tapestry5/jpa/test/JpaTest.java

```diff
@@ -0,0 +1,187 @@
+{
+
+	private static Registry registry;
+	private EntityManagerManager entityManagerManager;
+	private TopLevelService topLevelService;
+
+    // @BeforeSuite
+	public final void setupRegistry()
+	{
+		RegistryBuilder builder = new RegistryBuilder();
+		builder.add(TapestryModule.class);
+		builder.add(JpaModule.class);
+		builder.add(TransactionalUnitsModule.class);
+		builder.add(JpaTestModule.class);
+
+		registry = builder.build();
+		// set PageTesterContext, otherwise T5 tries to load classpath assets
+		ApplicationGlobals globals = registry.getObject(ApplicationGlobals.class, null);
+		globals.storeContext(new PageTesterContext(""));
+		registry.performRegistryStartup();
+
+		entityManagerManager = registry.getService(EntityManagerManager.class);
+		topLevelService = registry.getService(TopLevelService.class);
+
+	}
+
+	private EntityManager getEntityManager() {
+		return entityManagerManager.getEntityManagers().values().iterator().next();
+	}
+
+    // @AfterSuite
+	public final void shutdownRegistry()
+	{
+        registry.cleanupThread();
+		registry.shutdown();
+		registry = null;
+	}
+
+	@BeforeMethod
+	public final void beginTransaction()
+	{
+        setupRegistry();
+		EntityTransaction tx = getEntityManager().getTransaction();
+		if (!tx.isActive()) tx.begin();
+	}
+
+	@AfterMethod
+	public void rollbackLastTransactionAndClean() throws SQLException {
+		EntityTransaction transaction = getEntityManager().getTransaction();
+		if (transaction.isActive()) transaction.rollback();
+		clearDatabase();
+        getEntityManager().clear();
+        shutdownRegistry();
+	}
+
+	// based on http://www.objectpartners.com/2010/11/09/unit-testing-your-persistence-tier-code/
+	private void clearDatabase() throws SQLException {
+		EntityManager em = getEntityManager();
+        em.clear();
+		EntityTransaction transaction = em.getTransaction();
+		if (!transaction.isActive()) transaction.begin();
+		Connection c = em.unwrap(Connection.class);
+		Statement s = c.createStatement();
+		s.execute("SET REFERENTIAL_INTEGRITY FALSE");
+		Set<String> tables = new HashSet<String>();
+		ResultSet rs = s.executeQuery("select table_name " + "from INFORMATION_SCHEMA.tables "
+			+ "where table_type='TABLE' and table_schema='PUBLIC'");
+		while (rs.next()) {
+			// if we don't skip over the sequence table, we'll start getting "The sequence table information is not complete"
+			// exceptions
+			if (!rs.getString(1).startsWith("DUAL_") && !rs.getString(1).equals("SEQUENCE")) {
+				tables.add(rs.getString(1));
+			}
+		}
+		rs.close();
+		for (String table : tables) {
+			s.executeUpdate("DELETE FROM " + table);
+		}
+		transaction.commit();
+		s.execute("SET REFERENTIAL_INTEGRITY TRUE");
+		s.close();
+	}
+
+	private <T> List<T> getInstances(final Class<T> type) {
+		EntityManager em = getEntityManager();
+		CriteriaBuilder qb = em.getCriteriaBuilder();
+		CriteriaQuery<T> query = qb.createQuery(type);
+		return em.createQuery(query).getResultList();
+	}
```

Comment:
this method fails on line `em.createQuery(query).getResultList()` with:
> java.lang.IllegalStateException: No criteria query roots were specified

I had to rewrite it like this:
``` java
    private <T> List<T> getInstances(final Class<T> type)
    {
        EntityManager em = getEntityManager();

        if (em.getTransaction().isActive())
        {
            //  There shouldn't be any active transactions at this point,
            //  otherwise below query might return uncommitted changes
            em.getTransaction().rollback();
            em.clear();
        }

        CriteriaBuilder qb = em.getCriteriaBuilder();
        CriteriaQuery<T> query = qb.createQuery(type);
        Root<T> root = query.from(type);
        query.select(root);

        return em.createQuery(query).getResultList();
    }
```

Note that transaction rollback is necessary if current transaction is active, otherwise assertions may be incorrect.